aiida_nanotech_empa/workflows/cp2k/fragment_separation.py

```diff
@@ -0,0 +1,349 @@
+        spec.input("structure",
+                   valid_type=StructureData,
+                   help="A molecule on a substrate.")
+
+        spec.input_namespace(
+            "fragments",
+            valid_type=orm.List,
+            help="Fragments of the system to be treated individually.")
+
+        # Charges of each fragment.
+        spec.input_namespace(
+            "charges",  # +1 means one electron removed
+            valid_type=orm.Int,
+            help=
+            "Charges of each fragment. No need to specify the total charge as it would be computed automatically."
+        )
+
+        # Total multiplicity of the system and of the fragments.
+        spec.input_namespace(
+            "multiplicities",
+            valid_type=orm.Int,
+            #default=lambda: orm.Int(0),
+            help=
+            "Multiplicity of each fragment. Use 'total' to specify the multiplicity of the whole system."
+        )
+
+        # Fixed atoms and magnetization per site defined for the whole system. Will be automatically split for the fragments.
+        spec.input(
+            "fixed_atoms",
+            valid_type=orm.Str,
+            #default=lambda: orm.List(list=[]),
+            help="Fixed atoms of the system.")
+
+        spec.input(
+            "magnetization_per_site",
+            valid_type=orm.List,
+            #default=lambda: orm.List(list=[]),
+            help="Magnetization per site.")
+
+        # Protocol that defines the simulation settings.
+        spec.input(
+            "protocol",
+            valid_type=orm.Str,
+            #default=lambda: orm.Str('standard'),
+            required=False,
+            help="Settings to run simulations with.")
+
+        spec.input_namespace(
+            "auxilary_dictionaries",
+            valid_type=orm.Dict,
+            help=
+            "Dictionaries to append to the CP2K input dictionary right before the job submission. "
+            "It is useful to add constraints and collective variables to the input."
+        )
+
+        spec.input_namespace(
+            "options",
+            valid_type=dict,
+            non_db=True,
+            required=False,
+            help=
+            "Define options for the cacluations: walltime, memory, CPUs, etc.")
+
+        # Outline.
+        spec.outline(cls.setup, cls.run_scfs, cls.run_geo_opts, cls.finalize)
+
+        # Dynamic outputs.
+        spec.outputs.dynamic = True
+
+        # TODO: double-check the exit codes.
+        # Exit code.
+        spec.exit_code(
+            380,
+            "ERROR_SUBSTRATE_NOT_SUPPORTED",
+            message="Specified substrate is not supported.",
+        )
+        spec.exit_code(
+            381,
+            "ERROR_STRUCTURE_ANALYSIS",
+            message="Structure analysis failed.",
+        )
+        spec.exit_code(
+            390,
+            "ERROR_TERMINATION",
+            message="One or more steps of the work chain failed.",
+        )
```

Comment:
I need to better rework the exit_codes to have them compatible with the official guidelines (if there are any).

Reply:
Here are the guidelines: https://aiida.readthedocs.io/projects/aiida-core/en/latest/topics/processes/usage.html#exit-code-conventions